src/drivers/rest/data.rs

```diff
@@ -0,0 +1,165 @@
+
+use lazy_static::lazy_static;
+use mavlink::{self, Message};
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct Temporal {
+    first_update: chrono::DateTime<chrono::Local>,
+    last_update: chrono::DateTime<chrono::Local>,
+    counter: i64,
+    frequency: f32,
+}
+
+impl Default for Temporal {
+    fn default() -> Self {
+        Self {
+            first_update: chrono::Local::now(),
+            last_update: chrono::Local::now(),
+            counter: 1,
+            frequency: 0.0,
+        }
+    }
+}
+
+impl Temporal {
+    fn update(&mut self) {
+        self.last_update = chrono::Local::now();
+        self.counter = self.counter.wrapping_add(1);
+        self.frequency =
+            (self.counter as f32) / ((self.last_update - self.first_update).num_seconds() as f32);
+    }
+}
+
+#[derive(Default, Clone, Debug, Deserialize, Serialize)]
+struct Status {
+    time: Temporal,
+}
+
+impl Status {
+    fn update(&mut self) -> &mut Self {
+        self.time.update();
+        self
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct MAVLinkMessage<T> {
+    pub header: mavlink::MavHeader,
+    pub message: T,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct MAVLinkMessageStatus {
+    message: mavlink::ardupilotmega::MavMessage,
+    status: Status,
+}
+
+impl MAVLinkMessageStatus {
+    fn update(&mut self, message: &MAVLinkMessage<mavlink::ardupilotmega::MavMessage>) {
+        self.message = message.message.clone();
+        self.status.update();
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct MAVLinkVehicleComponentData {
+    id: u8,
+    messages: HashMap<String, MAVLinkMessageStatus>,
+}
+
+impl MAVLinkVehicleComponentData {
+    fn update(&mut self, message: &MAVLinkMessage<mavlink::ardupilotmega::MavMessage>) {
+        let message_name = message.message.message_name().to_string();
+        self.messages
+            .entry(message_name)
+            .or_insert(MAVLinkMessageStatus {
+                message: message.message.clone(),
+                status: Status::default(),
+            })
+            .update(message);
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct MAVLinkVehicleData {
+    id: u8,
+    components: HashMap<u8, MAVLinkVehicleComponentData>,
+}
+
+impl MAVLinkVehicleData {
+    fn update(&mut self, message: &MAVLinkMessage<mavlink::ardupilotmega::MavMessage>) {
+        let component_id = message.header.component_id;
+        self.components
+            .entry(component_id)
+            .or_insert(MAVLinkVehicleComponentData {
+                id: component_id,
+                messages: HashMap::new(),
+            })
+            .update(message);
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+pub struct MAVLinkVehiclesData {
+    vehicles: HashMap<u8, MAVLinkVehicleData>,
+}
+
+impl MAVLinkVehiclesData {
+    //TODO: Move message to reference
```

Comment:
I think it should be the opposite. Because that struct cannot be copied, to avoid the clone, we should not borrow, but move the ownership to all the update functions, and then you don't need to .clone() in the end to update the MAVLinkMessageStatus

Reply:
good point